### ISSUE TITLE
Not able to upgrade the content toolkit

### ISSUE THREAD

**Issue by brindha-mec** [oracle/content-and-experience-toolkit#35]
Hi, 

Downloaded the latest content toolkit and when running npm install, getting the below error -

127 verbose stack FetchError: request to https://registry.npmjs.org/@oracle%2Foraclejet failed, reason: unable to get local issuer certificate
127 verbose stack     at ClientRequest.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\minipass-fetch\lib\index.js:130:14)
127 verbose stack     at ClientRequest.emit (node:events:514:28)
127 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:501:9)
127 verbose stack     at TLSSocket.emit (node:events:526:35)
127 verbose stack     at emitErrorNT (node:internal/streams/destroy:151:8)
127 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
127 verbose stack     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
128 verbose cwd C:\content-and-experience-toolkit-master\sites
129 verbose Windows_NT 10.0.19045
130 verbose node v18.17.0
131 verbose npm  v9.6.7
132 error code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
133 error errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
134 error request to https://registry.npmjs.org/@oracle%2Foraclejet failed, reason: unable to get local issuer certificate
135 verbose exit 1

As mentioned in the logs, this is the node version : v18.17.0 and npm version:   v9.6.7
Also, downloaded the certificate from this URL : https://registry.npmjs.org/@oracle%2Foraclejet  and added to the cafile property and also tried adding the NODE_EXTRA_CA_CERTS value in the System environmental variables. Most of the forums suggesting to make the strict-ssl as false; I cannot do this as per our company policy. Also, should I be using any other node/npm versions? Anyone has come across this kind of error before? please advise.

Thanks!

**Comment by hzeng99** [comment]
could you check if the certificate is used?

npm config get cafile

**Comment by brindha-mec** [comment]
Yes, I manually downloaded the certificate from the browser and added the path -

 "user" config :

cafile = "C:\\Users\\Downloads\\oraclejet.pem"
progress = false
registry = "http://registry.npmjs.org/"

; "project" config from C:\content-and-experience-toolkit-master\sites\.npmrc

loglevel = "error"

; "cli" config from command line options



C:\content-and-experience-toolkit-master\sites>npm config get cafile
C:\Users\Downloads\oraclejet.pem

**Comment by hzeng99** [comment]
Your Node version should be fine. Is this the first time you trying to install toolkit?

**Comment by brindha-mec** [comment]
No, I have done many times. this is the first time with this node version.. :) 

I deleted all the contents from npm_cache and uninstalled and reinstalled nodejs, content toolkit. Also, added "NODE_EXTRA_CA_CERTS" system var and pointed this to the .pem file and restarted my system. Ran npm install and it's been more than an hour; BUT I got past this error message - I got a 200 in this step:

http fetch GET 200 https://registry.npmjs.org/@oracle%2Foraclejet 70595ms (cache stale)

I'll keep you posted :) I think it will take some more time; it is still running.

**Comment by brindha-mec** [comment]
it failed after 2 and half hours : 

https://registry.npmjs.org/mime failed, reason: unable to get local issuer certificate

2552 verbose stack FetchError: request to https://registry.npmjs.org/mime failed, reason: unable to get local issuer certificate
2552 verbose stack     at ClientRequest.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\minipass-fetch\lib\index.js:130:14)
2552 verbose stack     at ClientRequest.emit (node:events:514:28)
2552 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:501:9)
2552 verbose stack     at TLSSocket.emit (node:events:526:35)
2552 verbose stack     at emitErrorNT (node:internal/streams/destroy:151:8)
2552 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
2552 verbose stack     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
2553 verbose cwd C:\content-and-experience-toolkit-master\sites
2554 verbose Windows_NT 10.0.19045
2555 verbose node v18.17.0
2556 verbose npm  v9.6.7
2557 error code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
2558 error errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
2559 error request to https://registry.npmjs.org/mime failed, reason: unable to get local issuer certificate

**Comment by hzeng99** [comment]
what are the last working toolkit version and node version? Maybe try with the previous node version?

**Comment by brindha-mec** [comment]
I got the same error message with the previous version of the node (with the new content toolkit). that is why updated the node/npm version... I'll try after clearing cache and restart.. it's weird that it is able to get the cert for https://registry.npmjs.org/@oraclejet but not https://registry.npmjs.org/mime.. its is the same domain; so, should be the same cert...

**Comment by hzeng99** [comment]
Toolkit 23.4.1 changed to use oraclejet 14.0.4. What's your previous toolkit version?

**Comment by brindha-mec** [comment]
my previous version is 23.4.1.. I am trying to downgrade my nodejs version and trying to run npm install.

**Comment by brindha-mec** [comment]
I tried downgrading to my previous working npm/node version : node v16.18.0 and  npm 8.19.2 and tested with these versions of content toolkit - 

content-and-experience-toolkit-23.4.1 to the most recent one. getting the same error :(

**Comment by brindha-mec** [comment]
126 verbose stack     at ClientRequest.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\minipass-fetch\lib\index.js:130:14)
126 verbose stack     at ClientRequest.emit (node:events:513:28)
126 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:494:9)
126 verbose stack     at TLSSocket.emit (node:events:525:35)
126 verbose stack     at emitErrorNT (node:internal/streams/destroy:157:8)
126 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:122:3)
126 verbose stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
127 verbose cwd C:\content-and-experience-toolkit-23.5.2\sites
128 verbose Windows_NT 10.0.19045
129 verbose node v16.18.0
130 verbose npm  v8.19.2
131 error code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
132 error errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
133 error request to https://registry.npmjs.org/@oracle%2Foraclejet failed, reason: unable to get local issuer certificate

**Comment by hzeng99** [comment]
seems not caused by toolkit nor Node, could your env settings changed?

**Comment by hzeng99** [comment]
Maybe you can delete this line 

   "@oracle/oraclejet": "14.0.4",

from package.json and run npm install. Toolkit CLI commands should not be affected. 
May affect local server testing. I will run some test.

**Comment by brindha-mec** [comment]
Thank you.. I'll try that...

**Comment by brindha-mec** [comment]
> seems not caused by toolkit nor Node, could your env settings changed?

No.. not that I'm aware of..

**Comment by brindha-mec** [comment]
c:\content-and-experience-toolkit-master\sites>npm install
npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! request to https://registry.npmjs.org/body-parser failed, reason: unable to get local issuer certificate


Now getting the same error on /body-parser..

**Comment by brindha-mec** [comment]
sorry, this is not resolved yet.

**Comment by hzeng99** [comment]
then this has to with your vpn, proxy, ssl setttings

**Comment by brindha-mec** [comment]
I tried with and without proxy and got the same errors (different places like - one time on @oraclejet and sometimes on /mime) but same error message). Could you please advise what should the variable NODE_EXTRA_CA_CERTS point to?

**Comment by brindha-mec** [comment]
tried without vpn. getting this error now : 

npm ERR! Require stack:
npm ERR! - C:\content-and-experience-toolkit-master\sites\node_modules\progress\index.js
npm ERR! - C:\content-and-experience-toolkit-master\sites\node_modules\puppeteer\lib\cjs\puppeteer\node\install.js
npm ERR! - C:\content-and-experience-toolkit-master\sites\node_modules\puppeteer\install.js
npm ERR!     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
npm ERR!     at Function.Module._load (node:internal/modules/cjs/loader:778:27)
npm ERR!     at Module.require (node:internal/modules/cjs/loader:999:19)
npm ERR!     at require (node:internal/modules/cjs/helpers:102:18)
npm ERR!     at Object.<anonymous> (C:\content-and-experience-toolkit-master\sites\node_modules\progress\index.js:1:18)
npm ERR!     at Module._compile (node:internal/modules/cjs/loader:1099:14)
npm ERR!     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
npm ERR!     at Module.load (node:internal/modules/cjs/loader:975:32)
npm ERR!     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
npm ERR!     at Module.require (node:internal/modules/cjs/loader:999:19) {
npm ERR!   code: 'MODULE_NOT_FOUND',
npm ERR!   requireStack: [
npm ERR!     'C:\\content-and-experience-toolkit-master\\sites\\node_modules\\progress\\index.js',
npm ERR!     'C:\\content-and-experience-toolkit-master\\sites\\node_modules\\puppeteer\\lib\\cjs\\puppeteer\\node\\install.js',
npm ERR!     'C:\\content-and-experience-toolkit-master\\sites\\node_modules\\puppeteer\\install.js'
npm ERR!   ]
npm ERR! }
npm ERR!
npm ERR! Node.js v17.9.1

**Comment by hzeng99** [comment]
maybe change to use a Node LTS version, such as 18.17.1

**Comment by brindha-mec** [comment]
yes, changed it to node 18.17.1 and IT WORKED! (looks like it :) ) -

C:\content-and-experience-toolkit-master\sites>npm install

> cec-sites-toolkit@23.7.2 postinstall
> gulp --gulpfile bin/gulpfile.js copy-libs && npm install bin/cec/ --no-save || npm install bin\cec\ --no-save

Access is denied.

added 44 packages, and audited 705 packages in 13s

No errors but got this warning - Access is denied to the bin/cec folder..

**Comment by brindha-mec** [comment]
Also, getting this error when running cec install : 

C:\OCMSource\cec-src>cec install
Wed, 16 Aug 2023 03:04:55 GMT
Content Toolkit 23.7.2
cec install
Node version v18.17.0
Project set up, config OEC server in C:\OCMSource\cec-src\cec.properties
Install dependencies from package.json:
[22:05:17] 'install-src' errored after 7.05 s
[22:05:17] Error: ENOENT: no such file or directory, scandir 'C:\content-and-experience-toolkit-master\sites\src\libs'
    at Object.readdirSync (node:fs:1527:3)
    at C:\content-and-experience-toolkit-master\sites\bin\gulpfile.js:178:17
    at taskWrapper (C:\content-and-experience-toolkit-master\sites\node_modules\undertaker\lib\set-task.js:13:15)
    at bound (node:domain:433:15)
    at runBound (node:domain:444:12)
    at asyncRunner (C:\content-and-experience-toolkit-master\sites\node_modules\async-done\index.js:55:18)
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)

**Comment by brindha-mec** [comment]
Looking at the npm log, this is the actual error stack trace: 

2765 timing command:install Completed in 337011ms
2766 verbose stack Error: command failed
2766 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\@npmcli\promise-spawn\lib\index.js:53:27)
2766 verbose stack     at ChildProcess.emit (node:events:514:28)
2766 verbose stack     at maybeClose (node:internal/child_process:1091:16)
2766 verbose stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
2767 verbose pkgid cec-sites-toolkit@23.7.2
2768 verbose cwd C:\content-and-experience-toolkit-master\sites\bin
2769 verbose Windows_NT 10.0.19045
2770 verbose node v18.17.0
2771 verbose npm  v9.6.7
2772 error code 4294963238
2773 error path C:\content-and-experience-toolkit-master\sites\bin
2774 error command failed
2775 error command C:\WINDOWS\system32\cmd.exe /d /s /c gulp --gulpfile bin/gulpfile.js copy-libs && npm install bin/cec/ --no-save || npm install bin\cec\ --no-save
2776 error [[90m22:45:13[39m] No gulpfile found
2777 verbose exit 4294963238
2778 timing npm Completed in 338972ms
2779 verbose unfinished npm timer reify 1692157325947
2780 verbose unfinished npm timer reify:build 1692157483367
2781 verbose unfinished npm timer build 1692157483368
2782 verbose unfinished npm timer build:links 1692157510151
2783 verbose unfinished npm timer build:run:postinstall 1692157510151
2784 verbose unfinished npm timer build:run:postinstall:../../content-and-experience-toolkit-master/sites/bin 1692157510151
2785 verbose code 4294963238

**Comment by brindha-mec** [comment]
I think got this working. It's weird that when running cec install, gulpfile.js expecting /src/libs folder in my toolkit path which is C:/content-and-experience-toolkit-master/sites.. I created an empty folder /src/libs in it and it worked.. and I got a warning saying "Dependencies not installed; run npm install". I ran npm install again and looks like it finished OK.

Tried running cec develop and got my localhost. I really hope all the issues are resolved now. 

Thank you so much @hzeng99  for all your help and time :)
